tools/crowtty/src/main.rs

```diff
@@ -286,50 +298,51 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
+        // at all.
         while let Some(pos) = carry.iter().position(|b| *b == 0) {
-            let new_chunk = carry.split_off(pos + 1);
-            if let Ok(used) = cobs::decode_in_place(&mut carry) {
-                const PORT_LEN: usize = 2;
-                if carry.len() >= PORT_LEN {
-                    let mut bytes = [0u8; PORT_LEN];
-                    let (port, remain) = carry[..used].split_at(2);
-                    bytes.copy_from_slice(port);
-                    let port = u16::from_le_bytes(bytes);
-
+            let remainder = carry.split_off(pos + 1);
+
+            // Success means we printed something more useful than "bad decode",
+            // even if the actual decoding failed
+            let mut success = false;
+            match OwnedPortChunk::decode(&carry) {
+                Ok(OwnedPortChunk { port, chunk }) => {
+                    success = true;
                     if let Some(hdl) = manager.workers.get_mut(&port) {
                         if verbose {
-                            println!("{} {dmux} {}B -> :{port}", tag.port(port), remain.len());
+                            println!("{} {dmux} {}B -> :{port}", tag.port(port), chunk.len());
                         }
-                        hdl.out.send(remain.to_vec()).ok();
+                        hdl.out.send(chunk.to_vec()).ok();
                     }
-                } else {
-                    // we decoded a single byte, which we can't demux (as
-                    // there's no port number).
-                    let byte = carry[0];
-
-                    // if the byte was 0, it was probably sent to terminate any
-                    // existing frame, but there was no frame; in that case, we
-                    // can just ignore it. if it wasn't a zero, print it,
-                    // because that's weird.
-                    if byte != 0 {
-                        println!(
-                            "{tag} {dmux} {err} a {byte:#02x} is loose in the cat room! (no port)"
-                        );
+                }
+                Err(SPError::CobsDecodeFailed) => {
+                    if let Ok(s) = std::str::from_utf8(&carry[..]) {
+                        success = true;
+                        for line in s.lines() {
+                            println!("{tag} {text} {line}");
+                        }
                     }
                 }
-            } else if let Ok(s) = std::str::from_utf8(&carry[..]) {
-                for line in s.lines() {
-                    println!("{tag} {text} {line}");
+                Err(SPError::MalformedFrame) => {
+                    success = true;
+
+                    // If the malformed frame is JUST a null terminator, this is probably
+                    // a "frame flush" event, like we are just about to panic.
+                    if carry != &[0x00] {
+                        println!("{tag} {dmux} {err} bonus data? {carry:#02x?}");
+                    }
```

Comment:
nit, not actually a huge deal: perhaps the rule that a single null byte is acceptable and should just be ignored could be encoded in the `sermux-proto` crate --- we could have it return a different error when decoding a single null byte?